# Sprint 6 Project

Summary: 

# Importing the Neccessary Libraries

Code Meaning:
1) Import requests library in order to parse data from a website.
2) Import pandas library in order to change csv files into dataframes.
3) Import bs4 from BeautifulSoup library in order to translate parsed data int JSON format.
4) Import numpy library in order to run mathematical equations.
5) Import Plotly Express library in order to display charts.
6) Import stats from the scipy library in order to run t-tests and calculate p-value.

In [1]:
import requests

import pandas as pd

from bs4 import BeautifulSoup

import numpy as np

import plotly.express as px

from scipy import stats as st

# Uploading the Dataframes

Code Meaning:
1) Save the website address from the Sprint 6 project as a string in order to work parse the data from the website.
2) Save the string under the variable URL in order to save the result.

In [2]:
URL = 'https://practicum-content.s3.us-west-1.amazonaws.com/data-analyst-eng/moved_chicago_weather_2017.html'

Code Meaning:
1) Using the get() function from the requests library, in order to send the requests for data from the web server from the website provided in the URL variable.
2) Assign the code to req in order to save the result.

In [3]:
req = requests.get(URL)


1) Using the BeautifulSoup function from the bs4 library, parse the data from the req variable into lxml text format.
2) Assign the code to soup in order to save the result.

In [4]:
soup = BeautifulSoup(req.text, 'lxml')

Code Meaning:
1) Use the find function in order to find the first occurence of the value 'table'.  
2) Set the attributes variable to a dictionary titled, "id": "weather_records".
3) Assign the code to table in order to save the result.

In [5]:
table = soup.find('table', attrs={"id": "weather_records"})


Code Meaning:
1) Use the text function in order return a text output from the headers in the parsed data.
2) Within the code use the find_all() function to find all the 'th'(headers) sections in the parsed data (lxml) in order to retrieve them as input for the string.
3) Assign the code to headers in order to save the result.

In [6]:
headers = [header.text for header in table.find_all('th')]

Code Meaning:
1) Create an empty list and name the variable rows.
2) Create a for loop in which for every row

In [7]:
rows = []
for row in table.find_all('tr'):
    cells = row.find_all('td')
    if len(cells) > 0:
        rows.append([cell.text for cell in cells])

Code Results:
1) Using the pd.Dataframe() function, create a dataframe using the previously created rows variables as the input for the rows and use the headers variables as the input for the columns.
2) Assign the code to weather_records in order to save the result.
3) Use the head() method in order to print out the first five rows of the weather_records dataframe.

Conclusion:
The weather_records dataframe was created successfully, but the columms need to renamed in lower case format.

In [8]:
weather_records = pd.DataFrame(rows, columns=headers)

weather_records.head()

Date and time Temperature       Description
0  2017-11-01 00:00:00     276.150     broken clouds
1  2017-11-01 01:00:00     275.700  scattered clouds
2  2017-11-01 02:00:00     275.610   overcast clouds
3  2017-11-01 03:00:00     275.350     broken clouds
4  2017-11-01 04:00:00     275.240     broken clouds

Code Meaning:
1) Use the rename() method along withe the columns parameter in roder to rename the columns in the weather_records dataframe appropriately.
2) Reassign the code to the variable weather_records in order to save the result.
3) Use the head() method in order print the first five rows of the weather_records dataframe.

Code Result:
The date_and_time, temperature and description columns were renamed appropriately.

In [9]:
weather_records = weather_records.rename(columns={'Date and time' : 'date_and_time', 'Temperature': 'temperature', 'Description': 'description'})

weather_records.head()

date_and_time temperature       description
0  2017-11-01 00:00:00     276.150     broken clouds
1  2017-11-01 01:00:00     275.700  scattered clouds
2  2017-11-01 02:00:00     275.610   overcast clouds
3  2017-11-01 03:00:00     275.350     broken clouds
4  2017-11-01 04:00:00     275.240     broken clouds

Code Meaning:
1) Use the try method in order to create a file path that uploads the dataframe in a non-local computer.
2) In the try method, use pd.read_csv() in order to upload the dataframe under the file path: '/datasets/project_sql_result_01.csv'.
3) Assign the code to company_trips in order to save the result and create the dataframe.
4) Use the except method in order to create a file path that uploads the dataframe in my local computer.
5) In the except method, use pd.read_csv() in order to upload the dataframe under the file path: 'moved_project_sql_result_01.csv'.
6) Assign the code to company_trips in order to save the result and create the dataframe.
7) Use the head() method in order to print out the first five lines of the company_trips dataframe.

Code Result:
The company_trips dataframe was created successfully.

In [10]:
try:
    company_trips = pd.read_csv('/datasets/project_sql_result_01.csv')
except:
    company_trips = pd.read_csv('moved_project_sql_result_01.csv')

company_trips.head()

company_name  trips_amount
0                        Flash Cab         19558
1        Taxi Affiliation Services         11422
2                 Medallion Leasin         10367
3                       Yellow Cab          9888
4  Taxi Affiliation Service Yellow          9299

Code Meaning:
1) Use the try method in order to create a file path that uploads the dataframe in a non-local computer.
2) In the try method, use pd.read_csv() in order to upload the dataframe under the file path: '/datasets/project_sql_result_04.csv'.
3) Assign the code to dropoff_location in order to save the result and create the dataframe.
4) Use the except method in order to create a file path that uploads the dataframe in my local computer.
5) In the except method, use pd.read_csv() in order to upload the dataframe under the file path: 'moved_project_sql_result_04.csv'.
6) Assign the code to dropoff_location in order to save the result and create the dataframe.
7) Use the head() method in order to print out the first five lines of the company_trips dataframe.

Code Result:
The dropoff_location dataframe was created successfully.

In [11]:
try:
    dropoff_location = pd.read_csv('/datasets/project_sql_result_04.csv')
except:
    dropoff_location = pd.read_csv('moved_project_sql_result_04.csv')

dropoff_location.head()

dropoff_location_name  average_trips
0                  Loop   10727.466667
1           River North    9523.666667
2         Streeterville    6664.666667
3             West Loop    5163.666667
4                O'Hare    2546.900000

Code Meaning:
1) Use the try method in order to create a file path that uploads the dataframe in a non-local computer.
2) In the try method, use pd.read_csv() in order to upload the dataframe under the file path: '/datasets/project_sql_result_07.csv'.
3) Assign the code to weather_conditions in order to save the result and create the dataframe.
4) Use the except method in order to create a file path that uploads the dataframe in my local computer.
5) In the except method, use pd.read_csv() in order to upload the dataframe under the file path: 'moved_project_sql_result_07 (2).csv'.
6) Assign the code to weather_conditions in order to save the result and create the dataframe.
7) Use the head() method in order to print out the first five lines of the company_trips dataframe.

Code Result:
The weather_conditions dataframe was created successfully.

In [12]:
try:
    weather_conditions = pd.read_csv('/datasets/project_sql_result_07.csv')
except:
    weather_conditions = pd.read_csv('moved_project_sql_result_07 (2).csv')

weather_conditions.head()

start_ts weather_conditions  duration_seconds
0  2017-11-25 16:00:00               Good            2410.0
1  2017-11-25 14:00:00               Good            1920.0
2  2017-11-25 12:00:00               Good            1543.0
3  2017-11-04 10:00:00               Good            2512.0
4  2017-11-11 07:00:00               Good            1440.0

# Checking Data Types

Code Meaning:
1) Use the info() method on weather_records dataframe in order to check the datatypes in each column.

Code Result:
The temperature and description columns have the correct datatypes, but the date_and_time column needs to be converted from object data type to datetime64[ns] datatype.

In [13]:
weather_records.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 697 entries, 0 to 696
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   date_and_time  697 non-null    object
 1   temperature    697 non-null    object
 2   description    697 non-null    object
dtypes: object(3)
memory usage: 16.5+ KB


Code Meaning:
1) Use the pd.to_datetime() function in order to convert the date_and_time column of the weather_records dataframe from object to datetime64[ns].
2) Set the format parameter to: '%Y-%m-%d %H:%M:%S', in order to display the date and time correctly.
3) Assign the code to weather_records['date_and_time'] in order to save the result.

In [14]:
weather_records['date_and_time'] = pd.to_datetime(weather_records['date_and_time'], format='%Y-%m-%d %H:%M:%S')

Code Meaning:
1) Use the info() method in order to recheck the datatypes in the columns of the weather_records dataframe.

Code Result:
1) The date_and_time column of the weather_records dataframe was converted to datetime64[ns] successfully.

In [15]:
weather_records.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 697 entries, 0 to 696
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   date_and_time  697 non-null    datetime64[ns]
 1   temperature    697 non-null    object        
 2   description    697 non-null    object        
dtypes: datetime64[ns](1), object(2)
memory usage: 16.5+ KB


Code Meaning:
1) Use the info() method in order to check the data types in the columns of the company_trips dataframe.

Code Result:
1) The columns in the company_name dataframe have the appropriate data types.

In [16]:
company_trips.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64 entries, 0 to 63
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   company_name  64 non-null     object
 1   trips_amount  64 non-null     int64 
dtypes: int64(1), object(1)
memory usage: 1.1+ KB


Code Meaning:
1) Use the info() method in order to check the data types in the columns of the dropoff_location dataframe.

Code Result:
1) The columns in the dropoff_location dataframe have the appropriate data types.

In [17]:
dropoff_location.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94 entries, 0 to 93
Data columns (total 2 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   dropoff_location_name  94 non-null     object 
 1   average_trips          94 non-null     float64
dtypes: float64(1), object(1)
memory usage: 1.6+ KB


Code Meaning:
1) Use the info() method on weather_conditions dataframe in order to check the datatypes in each column.

Code Result:
The temperature and description columns have the correct datatypes, but the start_ts column needs to be converted from object data type to datetime64[ns] datatype.

In [18]:
weather_conditions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1068 entries, 0 to 1067
Data columns (total 3 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   start_ts            1068 non-null   object 
 1   weather_conditions  1068 non-null   object 
 2   duration_seconds    1068 non-null   float64
dtypes: float64(1), object(2)
memory usage: 25.2+ KB


Code Meaning:
1) Use the pd.to_datetime() function in order to convert the start_ts column of the weather_conditions dataframe from object to datetime64[ns].
2) Set the format parameter to: '%Y-%m-%d %H:%M:%S', in order to display the date and time correctly.
3) Assign the code to weather_conditions['start_ts'] in order to save the result.

In [19]:
weather_conditions['start_ts'] = pd.to_datetime(weather_conditions['start_ts'], format='%Y-%m-%d %H:%M:%S')

Code Meaning:
1) Use the info() method in order to recheck the datatypes in the columns of the weather_conditions dataframe.

Code Result:
1) The start_ts column of the weather_conditions dataframe was converted to datetime64[ns] successfully.

In [20]:
weather_conditions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1068 entries, 0 to 1067
Data columns (total 3 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   start_ts            1068 non-null   datetime64[ns]
 1   weather_conditions  1068 non-null   object        
 2   duration_seconds    1068 non-null   float64       
dtypes: datetime64[ns](1), float64(1), object(1)
memory usage: 25.2+ KB


# Checking For Duplicates

Code Meaning:
1) Use the duplicated() and sum() methods in order to count duplicate values in the weather_records dataframe.

Code Result:
There are no duplicate values noted in the weather_records dataframe.

In [21]:
weather_records.duplicated().sum()

0

Code Meaning:
1) Use the duplicated() and sum() methods in order to count duplicate values in the company_trips dataframe.

Code Result:
There are no duplicate values noted in the company_trips dataframe.

In [22]:
company_trips.duplicated().sum()

0

Code Meaning:
1) Use the duplicated() and sum() methods in order to count duplicate values in the dropoff_location dataframe.

Code Result:
There are no duplicate values noted in the dropoff_location dataframe.

In [23]:
dropoff_location.duplicated().sum()

0

Code Meaning:
1) Use the duplicated() and sum() methods in order to count duplicate values in the weather_conditions dataframe.

Code Result:
There are 197 duplicate values noted in the weather_conditions dataframe.

In [24]:
weather_conditions.duplicated().sum()

197

Code Meaning:
1) Use the drop_duplicates method in order to drop the duplicate values from the weather_conditions dataframe.
2) Set the inplace parameter to True in order to save the result.
3) Use the duplicated() and the sum() methods in order to count the number of duplicate values in the weather_conditions column once again.

Code Result:
There are no more duplicate values noted in the weather_conditions dataframe.

In [25]:
weather_conditions.drop_duplicates(inplace=True)

weather_conditions.duplicated().sum()

0

# Checking For Missing Values

Code Meaning:
1) Use the isna() and sum() methods in order to count the number of missing values in each column of the weather_records dataframe.

Code Result:
There are no missing values noted in the weather_records dataframe.

In [26]:
weather_records.isna().sum()

date_and_time    0
temperature      0
description      0
dtype: int64

Code Meaning:
1) Use the isna() and sum() methods in order to count the number of missing values in each column of the company_trips dataframe.

Code Result:
There are no missing values noted in the company_trips dataframe.

In [27]:
company_trips.isna().sum()

company_name    0
trips_amount    0
dtype: int64

Code Meaning:
1) Use the isna() and sum() methods in order to count the number of missing values in each column of the dropoff_location dataframe.

Code Result:
There are no missing values noted in the dropoff_location dataframe.

In [28]:
dropoff_location.isna().sum()

dropoff_location_name    0
average_trips            0
dtype: int64

Code Meaning:
1) Use the isna() and sum() methods in order to count the number of missing values in each column of the weather_conditions dataframe.

Code Result:
There are no missing values noted in the weather_conditions dataframe.

In [29]:
weather_conditions.isna().sum()

start_ts              0
weather_conditions    0
duration_seconds      0
dtype: int64

# Exploratory Data Analysis

### Bar Chart For Number of Rides Per Taxi Company (Task from Step 4)

Code Meaning:
1) Use the groupby() method on the company_trips dataframe.  Group the result by the company_name column.
2) Use the sum() function in order to sum up all the stacked values in the trips_amount column of the company_trips dataframe.
3) Assign the result to c_sum in order to save the result and to create a new dataframe.
4) Use the reset_index() method on the c_sum dataframe in order to reset the index.  
5) Use the rename() method and the columns parameter in order to rename the trips_amount column to trips_amount_sum().
6) Reassign the code to c_sum in order to save the result.
7) Use the sort_values() method in order to sort the column by  the values in the trips_amount_sum column.
8) Set the ascending parameter to False in order to sort the values in descending order.
9) Use the head() method and use 10 as an arguement in order to print the first 5 rows of the dataframe.

Code Result:
The c_sum dataframe was created successfully.

In [30]:
c_sum = company_trips.groupby('company_name')['trips_amount'].sum()
c_sum = c_sum.reset_index().rename(columns={'trips_amount': 'trips_amount_sum'})
c_sum = c_sum.sort_values(by='trips_amount_sum', ascending=False)

c_sum.head()

company_name  trips_amount_sum
45                        Flash Cab             19558
61        Taxi Affiliation Services             11422
50                 Medallion Leasin             10367
63                       Yellow Cab              9888
60  Taxi Affiliation Service Yellow              9299

Code Meaning:
1) Create a variable named company with the strings: 'Flash Cab', 'Taxi Affiliation Services', 'Medallion Leasin', 'Yellow Cab', 'Taxi Affiliation Service Yellow', 'Chicago Carriage Cab Corp', 'City Service', 'Sun Taxi', 'Star North Management LLC', 'Blue Ribbon Taxi Association Inc.'.  These will be the list of company names that will be used in the bar chart.
2) Use the isin() method on the company_name column of the c_sum dataframe with company as an arguement to check for presence values from company variable in the company_name column of the c_sum dataframe.
3) Assign the code to c_trip in order to save the result.
4) Place closed brackets [] next to the c_sum dataframe.  Insert the c_trip variable in the closed brackets in order to create a new dataframe.
5) Assign the code to c_trips in order to save the result and create a new dataframe.
6) Use the head() method in order to print the first five rows of the c_trips dataframe.

Code Result:
The c_trips dataframe was created successfully.

In [31]:
company = 'Flash Cab', 'Taxi Affiliation Services', 'Medallion Leasin', 'Yellow Cab', 'Taxi Affiliation Service Yellow', 'Chicago Carriage Cab Corp', 'City Service', 'Sun Taxi', 'Star North Management LLC', 'Blue Ribbon Taxi Association Inc.'

c_trip = c_sum['company_name'].isin(company)

c_trips = c_sum[c_trip]

c_trips.head()

company_name  trips_amount_sum
45                        Flash Cab             19558
61        Taxi Affiliation Services             11422
50                 Medallion Leasin             10367
63                       Yellow Cab              9888
60  Taxi Affiliation Service Yellow              9299

Code Meaning:
1) Use px.bar() in order to create a bar chart named company_bar using the data from the c_trips dataframe.
2) Set the x axis of the bar chart to company_name column of the c_trips dataframe and the y axis to the trips_amount_sum column.
3) Set the color parameter to the company_name column in order to color code the values along the x axis.
4) Use the labels parameter in order to labels parameter in order to relabel the x and y axes appropriately.
5) Set the title of the bar chart as: "Number of Rides Per Taxi Company".
6) Use the show() method in order to display the company_bar bar chart.

Code Result:
The company_bar bar chart was created successfully.

In [32]:
company_bar = px.bar(c_trips, x='company_name', y='trips_amount_sum', color='company_name', labels={'company_name': 'Company Name', 'trips_amount_sum': 'Trips Amount Sum'}, title='Number of Rides Per Taxi Company')
company_bar.show()

Conclusion/ Answer:
I chose 10 companies for this analysis which are:
1) Flash Cab- with an average trips amount sum of 19,558.
2) Taxi Affiliation Services- with an average trips amount sum of 11,422.
3) Medallion Leasin- with an average trips amount sum of 10,367.
4) Yellow Cab- with an average trips amount sum of 9,888.
5) Tax Affiliation Service Yellow- with an average trips amount sum of 9,299.
6) Chicago Carriage Cab Corp- with an average trips amount sum of 9,181.
7) City Service- with an average trips amount sum of 8,448.
8) Sun Taxi- with an average trips amount sum of 7,701.
9) Star North Management LLC- with an average trips amount sum of 7,455.
10) Blue Ribbon Taxi Association Inc.- with an average trips amount sum of 5,953.

### Identify the top 10 neighborhoods in terms of drop-offs (Task from Step 4)

Code Meaning:
1) Use the groupby() method on the dropoff_location dataframe.  Group the result by the dropoff_location_name column.
2) Use the sum() function in order to sum up all the stacked values in the average_trips column of the dropoff_location_name dataframe.
3) Assign the result to top_n_sum in order to save the result and to create a new dataframe.
4) Use the reset_index() method on the top_n_sum dataframe in order to reset the index.  
5) Use the rename() method and the columns parameter in order to rename the average_trips column to average_trips_sum().
6) Reassign the code to top_n_sum in order to save the result.
7) Use the sort_values() method in order to sort the column by  the values in the average_trips_sum column.
8) Set the ascending parameter to False in order to sort the values in descending order.
9) Use the head() method and use 10 as an arguement in order to print the first 10 rows of the dataframe.

Code Result:
The top_n_sum dataframe was created successfully.

In [33]:
top_n_sum = dropoff_location.groupby('dropoff_location_name')['average_trips'].sum()
top_n_sum = top_n_sum.reset_index().rename(columns={'average_trips': 'average_trips_sum'})
top_n_sum = top_n_sum.sort_values(by='average_trips_sum', ascending=False)
top_n_sum.head(10)

dropoff_location_name  average_trips_sum
50                  Loop       10727.466667
68           River North        9523.666667
78         Streeterville        6664.666667
87             West Loop        5163.666667
63                O'Hare        2546.900000
44             Lake View        2420.966667
35            Grant Park        2068.533333
56         Museum Campus        1510.000000
32            Gold Coast        1364.233333
74    Sheffield & DePaul        1259.766667

Code Meaning:
1) Create a variable named location with the strings: 'Loop', 'River North', 'Streeterville', 'West Loop', "O'Hare", 'Lake View', 'Grant Park', 'Museum Campus', 'Gold Coast', 'Sheffield & DePaul'.  These will be the list of dropoff location names that will be used in the bar chart.
2) Use the isin() method on the dropoff_location_name column of the top_n_sum dataframe with location as an arguement to check for presence values from location variable in the dropoff_location_name column of the top_n_sum dataframe.
3) Assign the code to top_10 in order to save the result.
4) Place closed brackets [] next to the top_n_sum dataframe.  Insert the location variable in the closed brackets in order to create a new dataframe.
5) Assign the code to top_10_df in order to save the result and create a new dataframe.
6) Use the head() method in order to print the first five rows of the top_10_df dataframe.

Code Result:
The top_10_df dataframe was created successfully.

In [34]:
location = 'Loop', 'River North', 'Streeterville', 'West Loop', "O'Hare", 'Lake View', 'Grant Park', 'Museum Campus', 'Gold Coast', 'Sheffield & DePaul'

top_10 = top_n_sum['dropoff_location_name'].isin(location)

top_10_df = top_n_sum[top_10]

top_10_df.head(10)


dropoff_location_name  average_trips_sum
50                  Loop       10727.466667
68           River North        9523.666667
78         Streeterville        6664.666667
87             West Loop        5163.666667
63                O'Hare        2546.900000
44             Lake View        2420.966667
35            Grant Park        2068.533333
56         Museum Campus        1510.000000
32            Gold Coast        1364.233333
74    Sheffield & DePaul        1259.766667

Code Meaning:
1) Use px.bar() in order to create a bar chart named top_10_bar using the data from the top_10_df dataframe.
2) Set the x axis of the bar chart to dropoff_location_name column of the top_10_df dataframe and the y axis to the average_trips_sum column.
3) Set the color parameter to the dropoff_location_name column in order to color code the values along the x axis.
4) Use the labels parameter in order to labels parameter in order to relabel the x and y axes appropriately.
5) Set the title of the bar chart as: "Top 10 Neighborhoods in Terms of Dropoffs".
6) Use the show() method in order to display the top_10_bar  bar chart.

Code Result:
The top_10_bar bar chart was created successfully.

In [35]:
top_10_bar = px.bar(top_10_df, x='dropoff_location_name', y='average_trips_sum', labels={'dropoff_location_name': 'Dropoff Location Name', 'average_trips_sum': 'Average Trips Sum'},title='Top 10 Neighborhoods in Terms of Dropoffs', color='dropoff_location_name')

top_10_bar.show()

Conclusion/ Answer:
The top 10 neighborhoods in terms of dropoffs are listed:  
1) Loop- with an average trips sum of 10,727.47.
2) River North- with an average trips sum of 9,523.
3) Streeterville- with an average trips sum of 6,664.667.
4) West Loop- with an average trips sum of 5,163.667.
5) O'Hare- with an average trips sum of 2,549.9.
6) Lake View- with an average trips sum of 2,420.
7) Grant Park- with an average trips sum of 2,068.533. 
8) Museum Campus- with an average trips sum of 1,510.
9) Gold Coast- with an average trips sum of 1,364.233.
10) Sheffield & DePaul- with an average trips sum of 1,259.767.

# Hypothesis Question (Task from Step 5)
"The average duration of rides from the Loop to O'Hare International Airport changes on rainy Saturdays."
Filter bad.
1 population is good and 1 population is bad and you test the hypothesis.
calculated the variance, calculate standard deviation. 

Null Hypothesis:
The average duration of rides from the Loop to O'Hare International Airport does not change on rainy Saturdays.

Alternate Hypothesis:
The average duration of rides from the Loop to O'Hare International Airport changes on rainy Saturdays.

Code Meaning:
1) Using conditional filtering, create a dataframe that only lists rows that have 'Good' as an entry for the weather_condiitons column.
2) Assign the code to good in order to save the result.

In [37]:
good = weather_conditions[weather_conditions['weather_conditions'] == 'Good']

Code Meaning:
1) Using conditional filtering, create a dataframe that only lists rows that have 'Bad' as an entry for the weather_condiitons column.
2) Assign the code to bad in order to save the result.

In [38]:
bad = weather_conditions[weather_conditions['weather_conditions'] == 'Bad']

Code Meaning:
1) Using the var() method from numpy, calculate the variance of the duration_seconds column of the bad dataframe.

Code Result:
The variance of the rows in the  duration_seconds column of the bad dataframe is 561003.0958729.

In [39]:
np.var(bad['duration_seconds'])

561003.0958729

Code Meaning:
1) Using the var() method from numpy, calculate the variance of the duration_seconds column of the good dataframe.

Code Result:
The variance of the rows in the  duration_seconds column of the good dataframe is 597389.5417434272.

Overall Conclusion:  The variances of the duration_seconds column of the bad dataframe and the good dataframe are not equal.  The variance of the durations_seconds column of the good dataframe is greater than the variance of the duration_seconds column of the bad dataframe.

In [40]:
np.var(good['duration_seconds'])

597389.5417434272

Code Meaning:
1) Using the std() method from numpy, calculate the standard deviation of the duration_seconds column of the bad dataframe.

Code Result:
The standard deviation of the rows in the  duration_seconds column of the bad dataframe is 749.00139991127787.

In [41]:
np.std(bad['duration_seconds'])

749.0013991127787

Code Meaning:
1) Using the std() method from numpy, calculate the standard deviation of the duration_seconds column of the good dataframe.

Code Result:
The standard deviation of the rows in the  duration_seconds column of the good dataframe is 772.9097888779952.

Overall Conclusion:
The standard deviation of the duration_seconds column of the good dataframe is greater than the standard deviation of the duration_seconds column of the bad dataframe.  The two standard deviations are not equal.

In [42]:
np.std(good['duration_seconds'])

772.9097888779952

Code Meaning:
1) Create a variable called alpha with the value 0.05.  This variable will represent the 5 percent probability that will be used for the ttest.
2) Use the ttest_ind() function from the scipy library with the duration_seconds column of the bad dataframe as the first arguement and the duration_seconds column of the good dataframe as the second arguement.  Set the equal_var parameter to False because the two variances are not equal.
3) Assign the code to the variable results in order to save the result.
4)  Print out the pvalue by using the code, results.pvalue.
5) Create an if-else statement that prints the string "We reject null hypothesis" if results.pvalue is less than alpha or prints the string "We cannot reject null hypothesis" if the else statement is valid. (results.pvalue >= alpha).

In [43]:
alpha = 0.05
results = st.ttest_ind(bad['duration_seconds'], good['duration_seconds'], equal_var=False)

print('pvalue:', results.pvalue)
if results.pvalue < alpha:
    print("We reject null hypothesis")
else:
    print("We cannot reject null hypothesis.")

pvalue: 9.126383933880083e-08
We reject null hypothesis


Conclusion/ Answer:
According to the ttest, the hypothesis that states: "The average duration of rides from the Loop to O'Hare International Airport changes on rainy Saturdays.", is correct because the null hypothesis has been rejected.

# Final Overall Conclusion:
